x-pack/plugins/apm/server/lib/transaction_groups/fetcher.ts

```diff
@@ -36,68 +44,149 @@ interface TopTraceOptions {
 export type Options = TopTransactionOptions | TopTraceOptions;
 
 export type ESResponse = PromiseReturnType<typeof transactionGroupsFetcher>;
+
+export type TransactionGroupRequestBase = ESSearchRequest & {
+  body: {
+    aggs: {
+      transaction_groups: Unionize<
+        Pick<AggregationOptionsByType, 'composite' | 'terms'>
+      >;
+    };
+  };
+};
+
+export type TransactionGroupSetup = Setup & SetupTimeRange & SetupUIFilters;
+
+function getItemsWithRelativeImpact(
+  setup: TransactionGroupSetup,
+  items: Array<{
+    sum?: number | null;
+    key: string | Record<string, any>;
+    avg?: number | null;
+    count?: number | null;
+    p95?: number;
+    sample?: Transaction;
+  }>
+) {
+  const values = items
+    .map(({ sum }) => sum)
+    .filter((value) => value !== null) as number[];
+
+  const max = Math.max(...values);
+  const min = Math.min(...values);
+
+  const duration = moment.duration(setup.end - setup.start);
+  const minutes = duration.asMinutes();
+
+  const itemsWithRelativeImpact: TransactionGroup[] = items
+    .map((item) => {
+      return {
+        key: item.key,
+        averageResponseTime: item.avg,
+        transactionsPerMinute: (item.count ?? 0) / minutes,
+        impact:
+          item.sum !== null && item.sum !== undefined
+            ? ((item.sum - min) / (max - min)) * 100 || 0
+            : 0,
+        p95: item.p95,
+        sample: item.sample!,
+      };
+    })
+    .filter((item) => item.sample);
```

Comment:
should the filter for `itemsWithRelativeImpact` be `.filter((item) => item.impact > 0)` or is that logically equivalent to whether or not there is a sample?

Reply:
I think it's pretty much the same, but the filter is mostly a safeguard for ensuring there's actually a sample as we can't deeplink without it. Previously a sample was guaranteed because it's all one request, but splitting it up into smaller requests means that is no longer the case - for instance, when there are more than 10k buckets.